sycl/include/sycl/detail/os_util.hpp

```diff
@@ -106,6 +106,24 @@ void fileTreeWalk(const std::string Path,
                   std::function<void(const std::string)> Func,
                   bool ignoreErrors = false);
 
+void *dynLookup(const char *WinName, const char *LinName, const char *FunName);
+
+// Look up a function name that was dynamically linked
+// This is used by the runtime where it needs to manipulate native handles (e.g.
+// retaining OpenCL handles). On Windows, the symbol name is looked up in
+// `WinName`. In Linux, it uses `LinName`.
+//
+// The library must already have been loaded (perhaps by UR), otherwise this
+// function throws a SYCL runtime exception.
+template <typename fn>
+fn *dynLookupFunction(const char *WinName, const char *LinName,
+                      const char *FunName) {
+  return reinterpret_cast<fn *>(dynLookup(WinName, LinName, FunName));
+}
+#define __SYCL_OCL_GET_FUNCTION(FN)                                            \
+  (sycl::_V1::detail::dynLookupFunction<decltype(FN)>("OpenCL",                \
+                                                      "libOpenCL.so", #FN))
```

Comment:
IMO, this generalization is unnecessary at this point. Can we just have a non-macro 
```
template <typename FnTy> FnTy *get_ocl_func(const char *FuncName) { ... }
void *get_ocl_func_impl(const char *FuncName);
```

? All the win/lin dispatch for the shared library name can be done inside `libsycl.so/.dll`

Reply:
Your `get_ocl_func` requires specifying the function name twice; once as a string and once as a value for `decltype(x)`. Without that, callers would have to do something like:

```c++
void *fn = reinterpret_cast<decltype(clDoThing)>("clDoThing");
```

which I think is less ergonomic than a macro.

`call_ocl` could work, but it looks like it relies on the programmer getting the signature exactly correct, and gives no feedback if they get it wrong.

Reply:
Yes, `__SYCL_OCL_CALL` looks much nicer. Updated.